apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4SamhsaMatcher.java

```diff
@@ -522,43 +523,89 @@ private static String normalizeIcd10Code(String icd10DiagnosisCode) {
   private boolean containsSamhsaProcedureCode(CodeableConcept procedureConcept) {
+    // Does the CodeableConcept have a legit HCPCS Coding?
+    boolean hasHcpcsCoding = findHcpcsCoding(procedureConcept);
+
+    // Check that Coding to see if it's blacklisted.
+    if (hasHcpcsCoding && isSamhsaCptCode(procedureConcept)) {
+      return true;
+    } else if (hasHcpcsCoding && hasUnknownSystem(procedureConcept)) {
+      /*
+       * Fail safe: if we don't know the procedure Coding system, assume the code is
+       * SAMHSA.
+       */
+      return true;
+    } else {
+      // Otherwise, this only has known & non-SAMHSA-blacklisted procedure codes.
+      return false;
+    }
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains the
+   *     {@link Coding} with the HCPCS {@link System}, <code>false</code> if it does not
+   */
+  private boolean findHcpcsCoding(CodeableConcept procedureConcept) {
     for (Coding procedureCoding : procedureConcept.getCoding()) {
       if (TransformerConstants.CODING_SYSTEM_HCPCS.equals(procedureCoding.getSystem())) {
-        if (isSamhsaCptCode(procedureCoding)) return true;
-      } else {
-        /*
-         * Fail safe: if we don't know the procedure Coding system, assume the code is
-         * SAMHSA.
-         */
         return true;
       }
     }
-
-    // No blacklisted procedure Codings found: this procedure isn't SAMHSA-related.
     return false;
   }
 
   /**
-   * @param procedureCoding the procedure {@link Coding} to check
-   * @return <code>true</code> if the specified procedure {@link Coding} matches one of the {@link
-   *     #cptCodes} entries, <code>false</code> if it does not
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>false</code> if the specified procedure {@link CodeableConcept} contains any
+   *     {@link Coding}s that do not contain any of the {@link System}s <code>true</code> if they do
    */
-  private boolean isSamhsaCptCode(Coding procedureCoding) {
+  private boolean hasUnknownSystem(CodeableConcept procedureConcept) {
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream()
+            .map(coding -> coding.getSystem())
+            .collect(Collectors.toSet());
+    // CHeck for HCPCS and HCPCS CD code systems, so that the right code is given in a backwards
+    // compatible way: https://jira.cms.gov/browse/BFD-1345
+    if (codingSystems.size() == 2
+        && codingSystems.contains(TransformerConstants.CODING_SYSTEM_HCPCS)
+        && codingSystems.contains(
+            TransformerUtilsV2.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD))) {
+      return false;
+    } else if (codingSystems.size() == 1
+        && codingSystems.contains(TransformerConstants.CODING_SYSTEM_HCPCS)) {
+      return false;
+    } else {
+      return true;
+    }
+  }
```

Comment:
Seems like maybe this could be condensed to:
```java
private boolean hasUnknownSystem(CodeableConcept concept) {
  Set<String> codingSystems =
        procedureConcept.getCoding().stream()
            .map(coding -> coding.getSystem())
            .collect(Collectors.toSet());
            
  return !codingSystems.isEmpty()
    && !codingSystems.equals(Set.of(TransformerConstants.CODING_SYSTEM_HCPCS))
    && !codingSystems.equals(Set.of(TransformerConstants.CODING_SYSTEM_HCPCS, CcwCodebookVariable.HCPCS_CD));
}```

Reply:
I agree that what is there now is a bit clunky. I am wondering if it might be more intuitive to write this function in positive logic and adjust the code that is calling it:

```java
private boolean hasKnownSystem(CodeableConcept concept) {
  Set<String> codingSystems =
        procedureConcept.getCoding().stream()
            .map(Coding::getSystem)
            .collect(Collectors.toSet());

  // Write a nice comment here about why this is considered to be a known system.
  Set<String> knownHcpcsSystem = Set.of(TransformerConstants.CODING_SYSTEM_HCPCS);

  // Write a nice comment here about why we need this.
  Set<String> backwardsCompatibleHcpcsSystem = Set.of(TransformerConstants.CODING_SYSTEM_HCPCS, TransformerUtilsV2.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD);
  
  return codingSystems.equals(knownHcpcsSystem) || codingSystem.equals(backwardsCompatibleHcpcsSystem);
}```

Reply:
One minor comment: I'm a bit concerned that there are V1/V2 versions of this method: `TransformerUtilsV2.calculateVariableReferenceUrl(...)`. If they're actually the same code, we should get rid of the duplicate. If they're _not_ the same code, it's probably a bad idea to use them here, as the SAMHSA filtering logic should not be version-specific.

Reply:
As an aside, that's delightfully clever code, @christrotteradhoc ! I haven't had enough coffee to try and verify it yet, but like I said, I'm going to mostly be watching the test coverage from here on out, anyways.

Reply:
@KyRobbins Same compliment for your proposed alternative, too. Bravo!

Reply:
> I agree that what is there now is a bit clunky. I am wondering if it might be more intuitive to write this function in positive logic and adjust the code that is calling it:
> 
> ```java
> private boolean hasKnownSystem(CodeableConcept concept) {
>   Set<String> codingSystems =
>         procedureConcept.getCoding().stream()
>             .map(Coding::getSystem)
>             .collect(Collectors.toSet());
> 
>   // Write a nice comment here about why this is considered to be a known system.
>   Set<String> knownHcpcsSystem = Set.of(TransformerConstants.CODING_SYSTEM_HCPCS);
> 
>   // Write a nice comment here about why we need this.
>   Set<String> backwardsCompatibleHcpcsSystem = Set.of(TransformerConstants.CODING_SYSTEM_HCPCS, TransformerUtilsV2.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD);
>   
>   return codingSystems.equals(knownHcpcsSystem) || codingSystem.equals(backwardsCompatibleHcpcsSystem);
> }
> ```

In that scenario, I think you'd want to name it `containsOnlyKnownSystems()`, as `hasKnownSystem()` infers that there is at least one known one.  Also not sure if you need to account for an empty list scenario.

Reply:
agreed and we ended up having to account for the empty systems in the higher level method to match v1 behavior.